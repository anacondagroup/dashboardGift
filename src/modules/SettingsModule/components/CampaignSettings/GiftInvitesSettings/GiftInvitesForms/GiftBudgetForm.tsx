import React, { useCallback, memo, useMemo } from 'react';
import { Box, InputAdornment, TextField, Typography } from '@mui/material';
import { ActionButton, AlyceTheme } from '@alycecom/ui';
import { TErrors } from '@alycecom/services';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { useExternalErrors } from '@alycecom/hooks';
import { Currencies, CurrencyConverter, ICurrency } from '@alycecom/modules';
import { makeStyles } from '@mui/styles';

import { ICampaignBudget } from '../../../../store/campaign/giftInvites/giftInvites.types';
import { campaignBudgetValidationSchema } from '../../../../store/campaign/giftInvites/giftInvites.schemas';
import { IPriceAvailability } from '../../../../../MarketplaceModule/store/priceAvailability/priceAvailability.types';
import GiftCardsMarketplaceLink from '../../../../../MarketplaceModule/components/Shared/GiftCardsMarketplaceLink';

export interface IBudgetFormProps {
  giftMinPrice: number;
  giftMaxPrice: number;
  giftCardPrice: number;
  giftDonationPrice: number;
  currencySign: string;
  campaignCurrencies: ICurrency[];
  priceAvailability: IPriceAvailability;
  countryIds: number[];
  onSave: (budget: ICampaignBudget) => void;
  isLoading?: boolean;
  errors?: TErrors<ICampaignBudget>;
}

const useStyles = makeStyles<AlyceTheme>(({ spacing }) => ({
  giftCardPreviewLink: {
    marginTop: spacing(1),
    marginBottom: spacing(4),
    textDecoration: 'none',
    '&:hover': {
      textDecoration: 'none',
    },
  },
}));

const GiftBudgetForm = ({
  giftMinPrice,
  giftMaxPrice,
  giftCardPrice,
  giftDonationPrice,
  currencySign,
  campaignCurrencies,
  priceAvailability,
  countryIds,
  onSave,
  isLoading = false,
  errors: externalErrors = {},
}: IBudgetFormProps): JSX.Element => {
  const classes = useStyles();
  const handleSaveBudget = useCallback(
    (data: ICampaignBudget) => {
      onSave({
        ...data,
        enterprise_min_price: data.enterprise_min_price ?? 0,
        enterprise_max_price:
          data.enterprise_max_price ?? (data.enterprise_gift_card_price > 1 ? data.enterprise_gift_card_price - 1 : 1),
        enterprise_gift_card_price: data.enterprise_gift_card_price ?? giftCardPrice,
        enterprise_donation_price: data.enterprise_donation_price ?? giftDonationPrice,
      });
    },
    [onSave, giftCardPrice, giftDonationPrice],
  );

  const {
    handleSubmit,
    setError,
    register,
    formState: { isValid, errors },
    watch,
  } = useForm<ICampaignBudget>({
    mode: 'all',
    defaultValues: {
      enterprise_min_price: giftMinPrice,
      enterprise_max_price: giftMaxPrice,
      enterprise_gift_card_price: giftCardPrice,
      enterprise_donation_price: giftDonationPrice,
    },
    resolver: yupResolver(campaignBudgetValidationSchema),
    context: { priceAvailability },
  });
  const { ref: minPriceRef, ...minPriceRegister } = register('enterprise_min_price');
  const { ref: maxPriceRef, ...maxPriceRegister } = register('enterprise_max_price');
  const { ref: cardPriceRef, ...cardPriceRegister } = register('enterprise_gift_card_price');
  const { ref: donationPriceRef, ...donationPriceRegister } = register('enterprise_donation_price');

  useExternalErrors<ICampaignBudget>(setError, externalErrors);

  const isSaveDisabled = isLoading || !isValid;

  const convertToCurrencies = useMemo(
    () =>
      campaignCurrencies.filter(currency => currency.id !== Currencies.constants.USD_ID).map(currency => currency.id),
    [campaignCurrencies],
  );
  const watchGiftCardPrice = watch('enterprise_gift_card_price');

  return (
    <form onSubmit={handleSubmit(handleSaveBudget)}>
      {priceAvailability.isPhysicalAvailable && (
        <Box mt={1}>
          <Typography className="Body-Regular-Left-Static-Bold">Merchandise</Typography>
        </Box>
      )}
      <Box display="flex" flexDirection="column">
        {priceAvailability.isPhysicalAvailable && (
          <Box width={360} display="flex" alignItems="flex-start" justifyContent="space-between">
            <Box width={160}>
              <TextField
                inputRef={minPriceRef}
                {...minPriceRegister}
                id="field-enterprise_min_price-id"
                type="number"
                name="enterprise_min_price"
                label="Min amount"
                placeholder="Min amount"
                fullWidth
                disabled={isLoading}
                variant="outlined"
                margin="normal"
                required
                InputProps={{
                  startAdornment: <InputAdornment position="start">{currencySign}</InputAdornment>,
                }}
                error={!!errors.enterprise_min_price}
                helperText={errors.enterprise_min_price && errors.enterprise_min_price.message}
              />
            </Box>
            <Box width={160}>
              <TextField
                inputRef={maxPriceRef}
                {...maxPriceRegister}
                id="field-enterprise_max_price-id"
                type="number"
                name="enterprise_max_price"
                label="Max amount"
                placeholder="Max amount"
                fullWidth
                disabled={isLoading}
                variant="outlined"
                margin="normal"
                required
                InputProps={{
                  startAdornment: <InputAdornment position="start">{currencySign}</InputAdornment>,
                }}
                error={!!errors.enterprise_max_price}
                helperText={errors.enterprise_max_price && errors.enterprise_max_price.message}
              />
            </Box>
          </Box>
        )}
        {priceAvailability.isDigitalAvailable && (
          <>
            <Box mt={3}>
              <Typography className="Body-Regular-Left-Static-Bold">Gift cards</Typography>
            </Box>
            <Box width={160}>
              <TextField
                inputRef={cardPriceRef}
                {...cardPriceRegister}
                id="field-enterprise_gift_card_price-id"
                type="number"
                name="enterprise_gift_card_price"
                label="Gift card price"
                placeholder="Gift card price"
                fullWidth
                disabled={isLoading}
                variant="outlined"
                margin="normal"
                required
                InputProps={{
                  startAdornment: <InputAdornment position="start">$</InputAdornment>,
                }}
                error={!!errors.enterprise_gift_card_price}
                helperText={errors.enterprise_gift_card_price && errors.enterprise_gift_card_price.message}
              />
            </Box>
            {convertToCurrencies.length > 0 && (
              <Box ml={6} width={350}>
                <CurrencyConverter.Component
                  price={Number(watchGiftCardPrice)}
                  fromCurrencyId={Currencies.constants.USD_ID}
                  toCurrencyIds={convertToCurrencies}
                />
              </Box>
            )}
            <GiftCardsMarketplaceLink
              className={classes.giftCardPreviewLink}
              giftCardPrice={watchGiftCardPrice}
              countryIds={countryIds}
            />
          </>
        )}
        {priceAvailability.isDonationAvailable && (
          <>
            <Box mt={1}>
              <Typography className="Body-Regular-Left-Static-Bold">
                Gift donation <span className="Body-Regular-Left-Static">- available in {currencySign} only</span>
              </Typography>
            </Box>
            <Box width={160}>
              <TextField
                inputRef={donationPriceRef}
                {...donationPriceRegister}
                id="field-enterprise_donation_price-id"
                type="number"
                name="enterprise_donation_price"
                label="Amount"
                placeholder="Amount"
                fullWidth
                disabled={isLoading}
                variant="outlined"
                margin="normal"
                required
                InputProps={{
                  startAdornment: <InputAdornment position="start">{currencySign}</InputAdornment>,
                }}
                error={!!errors.enterprise_donation_price}
                helperText={errors.enterprise_donation_price && errors.enterprise_donation_price.message}
              />
            </Box>
          </>
        )}
        <Box width={1} display="flex" justifyContent="space-between" mt={2}>
          <ActionButton type="submit" width={100} disabled={isSaveDisabled}>
            Save
          </ActionButton>
        </Box>
      </Box>
    </form>
  );
};

export default memo(GiftBudgetForm);
