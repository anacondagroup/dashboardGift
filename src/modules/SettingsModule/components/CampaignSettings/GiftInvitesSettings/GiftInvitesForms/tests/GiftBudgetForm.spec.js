import React from 'react';
import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';

import GiftBudgetForm from '../GiftBudgetForm';

jest.mock('react-router-dom', () => ({
  withRouter: jest.fn(),
  Link: 'Link',
}));
jest.mock('@mui/material', () => ({
  ...jest.requireActual('@mui/material'),
  TextField: 'TextField',
  Button: 'Button',
  Grid: 'Grid',
  Box: 'Box',
  InputAdornment: 'InputAdornment',
}));
jest.mock('@mui/styles', () => ({
  ...jest.requireActual('@mui/styles'),
  // eslint-disable-next-line global-require
  makeStyles: require('../../../../../../../libs/mocks').mockMakeStyles,
  withStyles: jest.fn(() => jest.fn()),
}));
jest.mock('react-redux', () => ({
  useSelector: jest.fn(() => jest.fn()),
  useDispatch: jest.fn(),
  connect: jest.fn(() => jest.fn()),
}));

describe('GiftBudgetForm', () => {
  let props;
  beforeEach(() => {
    props = {
      isLoading: false,
      giftMinPrice: 50,
      giftMaxPrice: 100,
      giftCardPrice: 100,
      giftDonationPrice: 100,
      onSave: () => {},
      errors: {},
      currencySign: '$',
      campaignCurrencies: [],
      priceAvailability: { isPhysicalAvailable: true, isDigitalAvailable: true, isDonationAvailable: true },
    };
  });

  it('should be rendered without errors', () => {
    const tree = create(shallow(<GiftBudgetForm {...props} />));
    expect(tree).toMatchSnapshot();
  });

  it('should be rendered with errors', () => {
    props.errors = {
      enterprise_min_price: ['Wrong min price'],
      enterprise_max_price: ['Wrong max price'],
      enterprise_gift_card_price: ['Wrong card price'],
      enterprise_donation_price: ['Wrong donation price'],
    };
    const tree = create(shallow(<GiftBudgetForm {...props} />));
    expect(tree).toMatchSnapshot();
  });

  it('should be show loading state', () => {
    props.isLoading = true;

    const tree = create(shallow(<GiftBudgetForm {...props} />));
    expect(tree).toMatchSnapshot();
  });
});
