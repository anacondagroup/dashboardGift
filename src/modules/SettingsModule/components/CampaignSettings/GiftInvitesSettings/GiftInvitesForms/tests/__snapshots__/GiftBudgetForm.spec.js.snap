// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GiftBudgetForm should be rendered with errors 1`] = `
<form
  onSubmit={[Function]}
>
  <Box
    mt={1}
  >
    <p
      className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
    >
      Merchandise
    </p>
  </Box>
  <Box
    display="flex"
    flexDirection="column"
  >
    <Box
      alignItems="flex-start"
      display="flex"
      justifyContent="space-between"
      width={360}
    >
      <Box
        width={160}
      >
        <TextField
          InputProps={
            Object {
              "startAdornment": <InputAdornment
                position="start"
              >
                $
              </InputAdornment>,
            }
          }
          disabled={false}
          error={false}
          fullWidth={true}
          id="field-enterprise_min_price-id"
          inputRef={[Function]}
          label="Min amount"
          margin="normal"
          name="enterprise_min_price"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Min amount"
          required={true}
          type="number"
          variant="outlined"
        />
      </Box>
      <Box
        width={160}
      >
        <TextField
          InputProps={
            Object {
              "startAdornment": <InputAdornment
                position="start"
              >
                $
              </InputAdornment>,
            }
          }
          disabled={false}
          error={false}
          fullWidth={true}
          id="field-enterprise_max_price-id"
          inputRef={[Function]}
          label="Max amount"
          margin="normal"
          name="enterprise_max_price"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Max amount"
          required={true}
          type="number"
          variant="outlined"
        />
      </Box>
    </Box>
    <Box
      mt={3}
    >
      <p
        className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
      >
        Gift cards
      </p>
    </Box>
    <Box
      width={160}
    >
      <TextField
        InputProps={
          Object {
            "startAdornment": <InputAdornment
              position="start"
            >
              $
            </InputAdornment>,
          }
        }
        disabled={false}
        error={false}
        fullWidth={true}
        id="field-enterprise_gift_card_price-id"
        inputRef={[Function]}
        label="Gift card price"
        margin="normal"
        name="enterprise_gift_card_price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Gift card price"
        required={true}
        type="number"
        variant="outlined"
      />
    </Box>
    <a
      className="MuiTypography-root MuiTypography-inherit local_giftCardPreviewLink MuiLink-root MuiLink-underlineAlways css-1iwtne7-MuiTypography-root-MuiLink-root"
      href="/marketplace/campaign?giftCardPrice=100&types=6"
      onBlur={[Function]}
      onFocus={[Function]}
      target="_blank"
    >
      Preview Gift Card Marketplace
    </a>
    <Box
      mt={1}
    >
      <p
        className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
      >
        Gift donation 
        <span
          className="Body-Regular-Left-Static"
        >
          - available in 
          $
           only
        </span>
      </p>
    </Box>
    <Box
      width={160}
    >
      <TextField
        InputProps={
          Object {
            "startAdornment": <InputAdornment
              position="start"
            >
              $
            </InputAdornment>,
          }
        }
        disabled={false}
        error={false}
        fullWidth={true}
        id="field-enterprise_donation_price-id"
        inputRef={[Function]}
        label="Amount"
        margin="normal"
        name="enterprise_donation_price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Amount"
        required={true}
        type="number"
        variant="outlined"
      />
    </Box>
    <Box
      display="flex"
      justifyContent="space-between"
      mt={2}
      width={1}
    >
      <Button
        className="local_button"
        disabled={true}
        type="submit"
      >
        Save
      </Button>
    </Box>
  </Box>
</form>
`;

exports[`GiftBudgetForm should be rendered without errors 1`] = `
<form
  onSubmit={[Function]}
>
  <Box
    mt={1}
  >
    <p
      className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
    >
      Merchandise
    </p>
  </Box>
  <Box
    display="flex"
    flexDirection="column"
  >
    <Box
      alignItems="flex-start"
      display="flex"
      justifyContent="space-between"
      width={360}
    >
      <Box
        width={160}
      >
        <TextField
          InputProps={
            Object {
              "startAdornment": <InputAdornment
                position="start"
              >
                $
              </InputAdornment>,
            }
          }
          disabled={false}
          error={false}
          fullWidth={true}
          id="field-enterprise_min_price-id"
          inputRef={[Function]}
          label="Min amount"
          margin="normal"
          name="enterprise_min_price"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Min amount"
          required={true}
          type="number"
          variant="outlined"
        />
      </Box>
      <Box
        width={160}
      >
        <TextField
          InputProps={
            Object {
              "startAdornment": <InputAdornment
                position="start"
              >
                $
              </InputAdornment>,
            }
          }
          disabled={false}
          error={false}
          fullWidth={true}
          id="field-enterprise_max_price-id"
          inputRef={[Function]}
          label="Max amount"
          margin="normal"
          name="enterprise_max_price"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Max amount"
          required={true}
          type="number"
          variant="outlined"
        />
      </Box>
    </Box>
    <Box
      mt={3}
    >
      <p
        className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
      >
        Gift cards
      </p>
    </Box>
    <Box
      width={160}
    >
      <TextField
        InputProps={
          Object {
            "startAdornment": <InputAdornment
              position="start"
            >
              $
            </InputAdornment>,
          }
        }
        disabled={false}
        error={false}
        fullWidth={true}
        id="field-enterprise_gift_card_price-id"
        inputRef={[Function]}
        label="Gift card price"
        margin="normal"
        name="enterprise_gift_card_price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Gift card price"
        required={true}
        type="number"
        variant="outlined"
      />
    </Box>
    <a
      className="MuiTypography-root MuiTypography-inherit local_giftCardPreviewLink MuiLink-root MuiLink-underlineAlways css-1iwtne7-MuiTypography-root-MuiLink-root"
      href="/marketplace/campaign?giftCardPrice=100&types=6"
      onBlur={[Function]}
      onFocus={[Function]}
      target="_blank"
    >
      Preview Gift Card Marketplace
    </a>
    <Box
      mt={1}
    >
      <p
        className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
      >
        Gift donation 
        <span
          className="Body-Regular-Left-Static"
        >
          - available in 
          $
           only
        </span>
      </p>
    </Box>
    <Box
      width={160}
    >
      <TextField
        InputProps={
          Object {
            "startAdornment": <InputAdornment
              position="start"
            >
              $
            </InputAdornment>,
          }
        }
        disabled={false}
        error={false}
        fullWidth={true}
        id="field-enterprise_donation_price-id"
        inputRef={[Function]}
        label="Amount"
        margin="normal"
        name="enterprise_donation_price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Amount"
        required={true}
        type="number"
        variant="outlined"
      />
    </Box>
    <Box
      display="flex"
      justifyContent="space-between"
      mt={2}
      width={1}
    >
      <Button
        className="local_button"
        disabled={true}
        type="submit"
      >
        Save
      </Button>
    </Box>
  </Box>
</form>
`;

exports[`GiftBudgetForm should be show loading state 1`] = `
<form
  onSubmit={[Function]}
>
  <Box
    mt={1}
  >
    <p
      className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
    >
      Merchandise
    </p>
  </Box>
  <Box
    display="flex"
    flexDirection="column"
  >
    <Box
      alignItems="flex-start"
      display="flex"
      justifyContent="space-between"
      width={360}
    >
      <Box
        width={160}
      >
        <TextField
          InputProps={
            Object {
              "startAdornment": <InputAdornment
                position="start"
              >
                $
              </InputAdornment>,
            }
          }
          disabled={true}
          error={false}
          fullWidth={true}
          id="field-enterprise_min_price-id"
          inputRef={[Function]}
          label="Min amount"
          margin="normal"
          name="enterprise_min_price"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Min amount"
          required={true}
          type="number"
          variant="outlined"
        />
      </Box>
      <Box
        width={160}
      >
        <TextField
          InputProps={
            Object {
              "startAdornment": <InputAdornment
                position="start"
              >
                $
              </InputAdornment>,
            }
          }
          disabled={true}
          error={false}
          fullWidth={true}
          id="field-enterprise_max_price-id"
          inputRef={[Function]}
          label="Max amount"
          margin="normal"
          name="enterprise_max_price"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Max amount"
          required={true}
          type="number"
          variant="outlined"
        />
      </Box>
    </Box>
    <Box
      mt={3}
    >
      <p
        className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
      >
        Gift cards
      </p>
    </Box>
    <Box
      width={160}
    >
      <TextField
        InputProps={
          Object {
            "startAdornment": <InputAdornment
              position="start"
            >
              $
            </InputAdornment>,
          }
        }
        disabled={true}
        error={false}
        fullWidth={true}
        id="field-enterprise_gift_card_price-id"
        inputRef={[Function]}
        label="Gift card price"
        margin="normal"
        name="enterprise_gift_card_price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Gift card price"
        required={true}
        type="number"
        variant="outlined"
      />
    </Box>
    <a
      className="MuiTypography-root MuiTypography-inherit local_giftCardPreviewLink MuiLink-root MuiLink-underlineAlways css-1iwtne7-MuiTypography-root-MuiLink-root"
      href="/marketplace/campaign?giftCardPrice=100&types=6"
      onBlur={[Function]}
      onFocus={[Function]}
      target="_blank"
    >
      Preview Gift Card Marketplace
    </a>
    <Box
      mt={1}
    >
      <p
        className="MuiTypography-root MuiTypography-body1 Body-Regular-Left-Static-Bold css-ahj2mt-MuiTypography-root"
      >
        Gift donation 
        <span
          className="Body-Regular-Left-Static"
        >
          - available in 
          $
           only
        </span>
      </p>
    </Box>
    <Box
      width={160}
    >
      <TextField
        InputProps={
          Object {
            "startAdornment": <InputAdornment
              position="start"
            >
              $
            </InputAdornment>,
          }
        }
        disabled={true}
        error={false}
        fullWidth={true}
        id="field-enterprise_donation_price-id"
        inputRef={[Function]}
        label="Amount"
        margin="normal"
        name="enterprise_donation_price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Amount"
        required={true}
        type="number"
        variant="outlined"
      />
    </Box>
    <Box
      display="flex"
      justifyContent="space-between"
      mt={2}
      width={1}
    >
      <Button
        className="local_button"
        disabled={true}
        type="submit"
      >
        Save
      </Button>
    </Box>
  </Box>
</form>
`;
